pilot/helpers/agents/CodeMonkey.py

```diff
@@ -1,38 +1,353 @@
+                logger.warning(f"No relevant files found for code change: {code_changes_description}")
+                return convo
+
+            # In practice, we always have exactly one file to change, so we can just pick the first in
+            # the list.
+            step['path'] = os.path.dirname(files_to_change[0])
+            step['name'] = os.path.basename(files_to_change[0])
+
+        rel_path, abs_path = self.project.get_full_file_path(step['path'], step['name'])
+
+        for f in files:
+            # Take into account that step path might start with "/"
+            if (f['path'] == step['path'] or (os.path.sep + f['path'] == step['path'])) and f['name'] == step['name'] and f['content']:
+                file_content = f['content']
+                break
+        else:
+            # If we didn't have the match (because of incorrect or double use of path separators or similar), fallback to directly loading the file
+            file_content = get_file_contents(abs_path, self.project.root_path)['content']
+            if isinstance(file_content, bytes):
+                file_content = "... <binary file, content omitted> ..."
+
+        file_name = os.path.join(rel_path, step['name'])
 
         llm_response = convo.send_message('development/implement_changes.prompt', {
-            "step_description": code_changes_description,
-            "step": step,
-            "task_description": task_description,
-            "step_index": step_index,  # todo remove step_index because in debugging we reset steps and it looks confusing in convo
-            "directory_tree": self.project.get_directory_tree(True),
-            "files": self.project.get_all_coded_files()  # self.project.get_files(files_needed),
-        }, IMPLEMENT_CHANGES)
-        convo.remove_last_x_messages(2)
+            "full_output": False,
+            "standalone": standalone,
+            "code_changes_description": code_changes_description,
+            "file_content": file_content,
+            "file_name": file_name,
+            "files": files,
+        })
+
+        exchanged_messages = 2
+        replace_complete_file = False
+
+        # Allow for up to 2 retries
+        while exchanged_messages < 7:
+            # Modify a copy of the content in case we need to retry
+            content = file_content
+
+            if re.findall('(old|existing).+code', llm_response, re.IGNORECASE):
+                llm_response = convo.send_message('utils/llm_response_error.prompt', {
+                    "error": (
+                        "You must not omit any code from NEW_CODE. "
+                        "Please don't use coments like `// .. existing code goes here`."
+                    )
+                })
+                exchanged_messages += 2
+                continue
+
+            # Split the response into pairs of old and new code blocks
+            blocks = self.get_code_blocks(llm_response)
```

Comment:
what if 2 files need to be changed but instead of 4 blocks we parse 2 blocks(wrong LLM response)? it would pass next 2 checks

Reply:
Aside from that fact that LLM would not us give complete file (for these two files that need to be changed)  because we asked it for old + new code blocks, it would also fail the third check because the new version of the first file would not be found in the old file.